### PR TITLE
Added DEmOS support for CPU benchmarks

### PR DESCRIPTION
**Pull request by MatejKafka**
Closes https://github.com/CTU-IIG/demos-sched/issues/28

### CODE REVIEW

**Comment by wentasah** [comment]
I'll look at that on Thursday. Now, I'm busy.

**Review by wentasah — CHANGES_REQUESTED** [review]
With respect to multi-threaded behavour, you're right. Can you define a global variable called something like `first_thread_id` and only call demos_completed if `thread_id == first_thread_id`? See other comments below.

File: benchmarks/CPU/instr/main.c [hunk 1]
```diff
@@ -44,7 +48,11 @@ void *benchmark_loop(void *ptr)
         }
         printf("CPU%d_work_done=%lu\n", thread_id, cpu_work_done);
         fflush(stdout);
-    }
+		// seems most sensible after printf, so that we see the progress before suspending
+		#ifdef DEMOS
+			demos_completed();
+		#endif
```
[the remark this answers is not in the document]
Comment (MatejKafka) [reply benchmarks/CPU/instr/main.c]: both ifdefs now also contain other code, it seems both simpler and cleaner to keep them than to add wrapper functions (and also confuse the reader as to why we're calling some random `demos_` functions without explaining that they're only invoked when built with demos support)

File: benchmarks/CPU/instr/main.c [hunk 2]
```diff
@@ -44,7 +48,11 @@ void *benchmark_loop(void *ptr)
         }
         printf("CPU%d_work_done=%lu\n", thread_id, cpu_work_done);
         fflush(stdout);
-    }
+		// seems most sensible after printf, so that we see the progress before suspending
```
[the remark this answers is not in the document]
Comment (MatejKafka) [reply benchmarks/CPU/instr/main.c]: fixed (was issue with tabs vs spaces)

**Comment by MatejKafka** [comment]
now behaves reasonably for multithreaded benchmarks

File: benchmarks/CPU/instr/main.c [hunk 3]
```diff
@@ -97,69 +96,64 @@ int main(int argc, char *argv[])
...
+    #endif
 
     for (int i = 0; i < num_proc; i++) {
         pthread_attr_t attr;
         cpu_set_t cpuset;
         pthread_t tid;
 
-        if ((cpu_mask & (1U << i)) == 0)
-            continue;
+        if ((cpu_mask & (1U << i)) == 0) continue;
```
Comment (wentasah) [inline benchmarks/CPU/instr/main.c:143]: No.

File: benchmarks/CPU/instr/main.c [hunk 4]
```diff
@@ -97,69 +96,64 @@ int main(int argc, char *argv[])
...
 
         CPU_ZERO(&cpuset);
         CPU_SET(i, &cpuset);
         pthread_attr_init(&attr);
         pthread_attr_setaffinity_np(&attr, sizeof(cpuset), &cpuset);
         int ret = pthread_create(&tid, &attr, benchmark_loop, (void *)(intptr_t)i);
-        if (ret != 0)
+        if (ret != 0) {
             warnx("Warning: CPU %d thread creation error: %s", i, strerror(ret));
+        }
```
Comment (wentasah) [inline benchmarks/CPU/instr/main.c:152]: I prefer not having extra braces if not necessary.

File: benchmarks/CPU/instr/meson.build [hunk 5]
```diff
@@ -64,21 +64,16 @@ demos_dep = meson.get_compiler('c').find_library(
...
 			output: 'read_demos_config.yaml',
-			command: ['cp', '@INPUT@', '@OUTPUT@'])]
-	demos_optional_dep = [demos_dep]
-	demos_flag = ['-DDEMOS=1']
-else
-	demos_sources = []
-	demos_optional_dep = []
-	demos_flag = []
+			command: ['cp', '@INPUT@', '@OUTPUT@'],
+			build_by_default: true)
```
Comment (wentasah) [inline benchmarks/CPU/instr/meson.build]: Just use
```meson
configure_file(copy: true,
	       input: 'read_demos_config.yaml',
	       output: 'read_demos_config.yaml')
```
it does exactly what's needed.

**Review by wentasah — CHANGES_REQUESTED** [review]
Few more changes is needed. And please, for future commits, start them with capital letters.

**Comment by wentasah** [comment]
Can you also rebase the PR to master? I've applied clang-format to all code and added github action to check formatting for all pushes and pull requests.

File: benchmarks/CPU/instr/main.c (hunk 3 above)
Comment (MatejKafka) [reply benchmarks/CPU/instr/main.c:143]: Current clang-format config formats the `continue;` to new line, which is imo 1) waste of screen space, 2) dangerous, as multiline ifs without braces are calling for trouble when someone adds statements after the if. I'll keep it the way it originally was for now, but for new code, I frankly don't think it's good style.

**Comment by MatejKafka** [comment]
Rebased on current master, cleaned up commit history, implemented most of the requested changes.

File: benchmarks/CPU/instr/main.c (hunk 3 above)
Comment (wentasah) [reply benchmarks/CPU/instr/main.c]: Don't discuss uselessly about style. It is an existing project which already uses some style. If we change the style to according to each contributor, it'd be a mess.

**Review by wentasah — APPROVED** [review]
I think the rebase was not successful. Your branch contains commits, which are already in master, but with different hashes. Never mind. I manually cherry picked your commits that I liked and ignored the formatting and comment-style changes. They do not add anything useful.

I also added documentation and a few other changes that make it easier to use the DEmOS support and pushed it to master. So I consider this as done.

**Pull request closed without merging**